### PR TITLE
feat(sandbox-runtime): list-native repo sync, hooks, manifest, push targeting

### PR DESCRIPTION
**Pull request by ColeMurray**
## Summary

PR-1 of the multi-repo sessions plan (`docs/multi-repo-sessions-implementation-plan.md`, design `docs/multi-repo-sessions-design.md` §6.6/§5.4/§6.5): the sandbox runtime becomes **list-native**. `SESSION_CONFIG.repositories` (ordered, first = primary) drives every code path; absent, a one-entry list is synthesized from `REPO_OWNER`/`REPO_NAME`, so single-repo sessions run the exact same code. **Inert until the control plane sends the new fields** (PR-2/PR-4).

### Runtime (`entrypoint.py`)
- **Unified per-repo sync rule** for every boot mode: existing checkout → set-url/fetch/checkout, missing → clone; members sync concurrently (`asyncio.gather`). Git primitives are parameterized by repo.
- **Failure policy split** (D§6.6): fresh/build boots are **fatal** on any member failure, naming the member (**deliberate behavior change** — previously a fresh boot limped on repo-less); image/snapshot boots keep their leniency and queue a `warning` instead (a deleted upstream branch must not brick a resume).
- **Hooks per member, position order**: setup only on fresh/build boots (build fatal on first failing member; fresh warns and continues); start on non-build boots (primary fatal, secondaries warn and continue).
- **OpenCode + code-server root**: `/workspace` when N>1 (or repo-less); the repo itself for single-repo sessions (unchanged).
- **Generated workspace manifest** (`/workspace/AGENTS.md`, multi-repo only, regenerated every boot): member table with base branches, the shared working-branch name read from `SESSION_CONFIG.working_branch_name` (computed control-plane-side — the runtime never derives branch names), per-member `AGENTS.md` enumeration, and the `create-pull-request repo:` instruction.
- **`.opencode` assembly** (multi-repo only): member repos' `.opencode/` merged into `/workspace/.opencode/` in position order, last-write-wins with a collision warning naming both members; system tools still install after (single-repo parity).
- **Boot warnings channel**: the supervisor has no control-plane event channel, so warnings append to `/tmp/oi-boot-warnings.jsonl` (cleared every boot); the bridge drains the file into `warning` sandbox events after its WebSocket handshake. The event schema member lands in PR-3 — until then live ingest drops them, harmlessly.

### Bridge (`bridge.py`)
- `_handle_push` targets the member named by `pushSpec.repoOwner/repoName`; unknown member → `push_error` naming it; specs without identity keep today's sole-clone behavior. `push_complete`/`push_error` echo repo identity when the spec carried it, and the "No repository found" path now includes `branchName` (today's key-less event is dropped at CP ingest, leaking the pending push resolver for 360s — PR-3/5 finish this fix).
- Git identity is configured in **every** member checkout (previously first-glob only).

### Contracts & plumbing
- `types.py` `SessionConfig` gains `repositories` + `working_branch_name` — load-bearing: this pydantic model is round-tripped by modal-infra and silently drops unknown keys.
- `inspect-plugin.js` `create-pull-request` gains an optional `repo` (`"owner/name"`) arg — required (with the valid list in the error) when the session spans multiple repositories; `headBranch` resolves via `git -C /workspace/{name}`; body forwards `repoOwner`/`repoName`. Also fixes the stale `baseBranch` doc (the handler already defaults to the **session** base branch — D§6.7-3 correction).
- `CACHE_BUSTER` → `v53-list-native-runtime` (the new runtime must be in the base image before any multi-repo session can boot).

### Tests
392 pass in sandbox-runtime (was 364 pre-PR), 179 in modal-infra. New coverage: repository-list parsing/synthesis; sync failure lists; fresh-fatal vs snapshot-warn policy (run()-level); hook ordering + full fatality matrix; workdir rule; manifest content; assembly order/collision/node_modules skip; boot-warning recorder + bridge drain (incl. malformed lines, exactly-once); push targeting/unknown-member/legacy-fallback/branch-on-no-repo; SessionConfig round-trip. Existing suites updated for the per-repo signatures.

Writing the tests surfaced a real bug in the first cut: passing `message` as a structured-log kwarg collides with Python logging's reserved LogRecord field and would have turned every boot warning into a fatal spawn error — now pinned by tests.

### Deploy note
Merge **after** PR #897 (phase order; no code dependency). Runtime changes are inert until the control plane sends `repositories` — single-repo boots exercise the synthesized one-entry path, covered by the full regression suite.

## Summary by CodeRabbit

* **New Features**
  * Added multi-repository session support across repository sync, per-repository setup/start hooks, and workspace preparation (including merged opencode assets).
  * Pull request creation can target a specific repository in multi-repo sessions, including repository identity in the request.
  * Multi-repo push now selects the correct checkout based on repo identity and reports identity details in success/errors.
* **Bug Fixes**
  * Boot warnings queued during startup are drained and emitted after the initial handshake.
  * Improved git commit attribution by applying user name/email to all member checkouts.
* **Tests**
  * Expanded coverage for multi-repo workspace, repo manifests, boot warnings, and push/error edge cases.

### CODE REVIEW

File: packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py [hunk 1]
```diff
@@ -154,13 +178,47 @@ def base_branch(self) -> str:
...
+                name = str(item.get("repo_name") or "").strip()
+                if not owner or not name:
+                    continue
+                branch = str(item.get("branch") or "").strip() or "main"
+                entries.append(
+                    RepoEntry(
+                        owner=owner,
+                        name=name,
+                        branch=branch,
+                        path=self.workspace_path / name,
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py]: Fixed in 7a281b24 along the lines you suggested. A new `repo_config` module now owns the model: `parse_repositories()` validates owner/name as safe single path segments, rejects duplicate names (case-insensitive), and computes the checkout path once on `RepoEntry`. The supervisor then writes a machine-readable manifest (`/tmp/oi-repo-manifest.json`, `REPO_MANIFEST_FILE_PATH`) that the bridge and the JS plugin consume — sync, hooks, manifest generation, push targeting, and the plugin all resolve paths through that one model instead of recomputing `workspace / name`.

[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py]: Fixed in 7a281b24: unsafe owner/name values (`..`, separators, absolute paths) and duplicate names (case-insensitive) now raise `RepoConfigError` at parse time. The error is deferred to `run()` so it reaches the control plane through the normal fatal-error path — the boot refuses to proceed rather than silently dropping or disambiguating members (the control plane rejects duplicate names at create time, so a violation here means a corrupt or tampered config). The scalar env fallback is validated by the same rule. Regression tests cover traversal, absolute paths, backslashes, and cross-owner duplicates.

File: packages/sandbox-runtime/src/sandbox_runtime/bridge.py [hunk 2]
```diff
@@ -1481,26 +1513,64 @@ async def _handle_push(self, cmd: dict[str, Any]) -> None:
...
+                fields["repoOwner"] = repo_owner
+            if repo_name:
+                fields["repoName"] = repo_name
+            return fields
+
+        # Resolve the target checkout. A spec carrying repo identity targets
+        # that member directly (multi-repo sessions); a spec without one keeps
+        # the legacy behavior — the sole cloned repository under /workspace.
+        if repo_name:
+            repo_dir = self.repo_path / repo_name
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/bridge.py]: Fixed in 7a281b24: the push path now resolves through the same canonical model as boot sync. If either `repoOwner` or `repoName` is present, both are required (partial specs get an explicit `push_error` instead of a silent fallback); the pair is matched case-insensitively against the supervisor-written manifest and the member's canonical path is used — spec-supplied names are never joined into the filesystem, so traversal values can't select a checkout. Non-members error without touching disk. Tests cover partial identity, non-members, `../` names, and case-insensitive canonical-path targeting.

[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/bridge.py]: Fixed in 7a281b24 — see the thread above: targeting now goes through the manifest membership check, so `repoName` is never used as a path segment at all.

File: packages/sandbox-runtime/src/sandbox_runtime/plugins/inspect-plugin.js [hunk 3]
```diff
@@ -45,9 +45,29 @@ function getSessionId() {
...
+      .filter((entry) => entry.owner && entry.name);
+  } catch (e) {
+    console.log("[create-pull-request] Failed to parse SESSION_CONFIG repositories:", e.message);
+    return [];
+  }
+}
+
+async function getCurrentBranch(repoName) {
+  try {
+    const gitArgs = repoName
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/plugins/inspect-plugin.js]: Fixed in 7a281b24: the plugin now reads the supervisor-generated workspace manifest (`/tmp/oi-repo-manifest.json`) instead of parsing SESSION_CONFIG and inferring `/workspace/${repoName}`. After the case-insensitive match it uses the matched entry's canonical owner/name for the bridge call and the entry's `path` for branch lookup — the checkout-path convention now lives only in the Python `repo_config` module that writes the manifest.

File: packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py [hunk 4]
```diff
@@ -428,31 +489,176 @@ async def _get_head_sha(self) -> str:
...
+        OpenCode discovers config relative to its cwd — /workspace for
+        multi-repo sessions — so per-repo custom tools/skills/commands would
+        never load. Files are copied in position order, last write wins with a
+        warning naming both members; the system tools installed afterwards
+        still override on filename collision (same as single-repo today).
+        """
+        if not self.is_multi_repo:
+            return
+
+        dest_root = self.workspace_path / ".opencode"
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py:558]: Fixed in 7a281b24: `_assemble_workspace_opencode()` now removes `/workspace/.opencode` before merging, so the tree is rebuilt from scratch on every boot. That's safe because assembly runs before `_install_tools`/`_stage_opencode_deps`/`_install_skills` in `_prepare_opencode_filesystem` — system tools and staged deps are re-installed after assembly each boot. Regression test: a stale generated file from a previous boot's member set is gone after reassembly.

**Comment by ColeMurray** [comment]
@coderabbitai Addressed your findings in 7a281b24:

- **Repo path collision on duplicate `name`**: rather than silently disambiguating with `owner-name` directories (which would diverge from the `/workspace/<name>` convention the control plane, AGENTS.md manifest, and PR tooling assume), duplicate names now fail the boot with a fatal `RepoConfigError`. The control plane rejects duplicate-name sessions at create time (`sessionRepositoriesInputSchema`), so the runtime treats a duplicate as a corrupt/tampered config.
- **Path-segment validation**: owner and name must match `^[A-Za-z0-9._-]+$` (excluding `.`/`..`) before any path is derived; the scalar env fallback is validated by the same rule.
- **Unhandled exception in `_clone_repo`**: the subprocess exec/communicate is now wrapped so an `OSError` surfaces as a failed member through `sync_repositories`' partial-failure contract instead of propagating through the gather.
- **Plugin canonical casing**: the matched manifest entry's owner/name/path are used after the case-insensitive lookup.

File: packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py:558]: Follow-up in ed71c5d4: the clean rebuild now spares `node_modules`. The blanket rmtree was forcing `_stage_opencode_deps` to re-copy the staged `@opencode-ai/plugin` tree on every multi-repo snapshot restore, defeating its skip-if-present fast path. Assembly never writes into `node_modules` (member `node_modules` are skipped during the merge), so the entry is purely image-managed and can't hold stale member state. `package.json`/`package-lock.json` are still deleted since a member can legitimately ship `.opencode/package.json`.
